File: examples/dart/lib/auth/usecase/jwt.dart
// lib/auth/usecase/jwt.dart — JWT token generation and validation

import '../domain/user.dart';
import '../domain/errors.dart';

class JwtService {
  Future<String> generateToken(User user) async {
    return 'fake-jwt-token-for-${user.id}';
  }

  Future<User?> validateToken(String token) async {
    // In a real app, this would decode and verify the JWT
    return null;
  }

  Future<User> requireAuth(String? token) async {
    if (token == null || token.isEmpty) {
      throw DomainError(code: 'unauthorized', message: 'Authorization header required');
    }
    final user = await validateToken(token);
    if (user == null) {
      throw NotFoundError('token');
    }
    return user;
  }
}

File: examples/typescript/src/auth/usecase/jwt.ts
// src/auth/usecase/jwt.ts — JWT token generation and validation

import { User, UserRepository } from "../domain/user";
import { TokenPayload, TokenRepository } from "../domain/token";
import { NotFoundError, DomainError } from "../domain/errors";

export class JwtService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly tokenRepo: TokenRepository,
  ) {}

  async generateToken(user: User): Promise<string> {
    const token = `fake-jwt-token-for-${user.id}`;
    const payload: TokenPayload = {
      sub: user.id,
      role: user.role,
      exp: Date.now() + 3600000,
    };
    await this.tokenRepo.save(token, payload);
    return token;
  }

  async validateToken(token: string): Promise<User | null> {
    const payload = await this.tokenRepo.findByToken(token);
    if (!payload) {
      return null;
    }
    const user = await this.userRepo.findById(payload.sub);
    return user;
  }

  async requireAuth(token: string | undefined): Promise<User> {
    if (!token) {
      throw new DomainError("unauthorized", "Authorization header required");
    }
    const user = await this.validateToken(token);
    if (!user) {
      throw new NotFoundError("token");
    }
    return user;
  }
}
